lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,594 @@
+
+import logging
+
+from ops.charm import CharmBase, InstallEvent
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus
+
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateRequestEvent,
+    TLSCertificatesProvides,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExampleProviderCharm(CharmBase):
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.certificates = TLSCertificatesProvides(self, "certificates")
+        self._stored.set_default(ca_private_key=b"", ca_certificate=b"")
+        self.framework.observe(
+            self.certificates.on.certificate_request, self._on_certificate_request
+        )
+        self.framework.observe(self.on.install, self._on_install)
+
+    def _on_install(self, event: InstallEvent) -> None:
+        self._stored.ca_private_key = generate_private_key()
+        self._stored.ca_certificate = generate_ca(
+            private_key=self._stored.ca_private_key, subject="whatever"
+        )
+
+    def _on_certificate_request(self, event: CertificateRequestEvent) -> None:
+        certificate = generate_certificate(
+            ca=self._stored.ca_certificate,
+            ca_key=self._stored.ca_private_key,
+            csr=event.certificate_signing_request.encode()
+        )
+
+        self.certificates.set_relation_certificate(
+            certificate=certificate.decode("utf-8"),
+            certificate_signing_request=event.certificate_signing_request,
+            ca=self._stored.ca_certificate.decode("utf-8"),
+            chain=self._stored.ca_certificate.decode("utf-8"),
+            relation_id=event.relation_id
+        )
+        self.unit.status = ActiveStatus()
+
+
+if __name__ == "__main__":
+    main(ExampleProviderCharm)
+```
+
+### Requirer charm
+Example:
+
+```python
+import logging
+
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateAvailableEvent,
+    TLSCertificatesRequires,
+    generate_private_key
+)
+from ops.framework import StoredState
+from ops.charm import CharmBase, RelationJoinedEvent
+from ops.model import ActiveStatus
+from ops.main import main
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExampleRequirerCharm(CharmBase):
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.certificates = TLSCertificatesRequires(self, "certificates")
+        self._stored.set_default(private_key=b"", private_key_password=b"")
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.on.certificates_relation_joined, self._on_certificates_relation_joined
+        )
+        self.framework.observe(self.on.install, self._on_install)
+
+    def _on_install(self, event) -> None:
+        self._stored.private_key_password = b"banana"
```

Comment:
An exquisite choice of passphrase. Bravo.

Reply:
Thank you :)